### PR TITLE
Refactored sprint and added pages for printing source plate labels, c…

### LINKED ISSUE #42: GPL-785 As developers we need to add ad hoc printing so that user can print off plate labels in the lab
**User story**
As developers we need to add ad hoc printing so that user can print off plate labels in the lab. At the moment we are having to print off labels ad hoc using scripts and it is time consuming.

**Who are the primary contacts for this story**
Steve

**Acceptance criteria**
To be considered successful the solution must allow:
- [ ] print off source plate labels specifying number of positives
- [ ] print off control plate labels

**Dependencies**
This story is blocked by the following dependencies:
- #<issue_no.>
- sanger/<repo>#<issue_no.>

**Additional context**

### PR DESCRIPTION
**Pull request by stevieing**
…ontrol plates and ad hoc printing.

Closes #42

Changes proposed in this pull request:

*
*
* ...

### CODE REVIEW

**Comment by stevieing** [comment]
I have deployed as is to uat via branch.

To print source labels:

visit /print_labels/source_plates. Upload a file in the format shown below as csv:

barcode,text
AP-RNA-26020001,number of positives: 10
AP-RNA-26020002,number of positives: 0
AP-RNA-26020003,number of positives: 20
AP-RNA-26020004,number of positives: 3
AP-RNA-26020005,number of positives: 10
AP-RNA-26020006,number of positives: 10
AP-RNA-26020007,number of positives: 10
AP-RNA-26020008,number of positives: 10
AP-RNA-26020009,number of positives: 10
AP-RNA-26020010,number of positives: 10

all you need is two columns with the headers as shown. It will need to be csv. It doesn't matter what is in the text column.

to print control plate labels:

visit /print_labels/control_plates

scan a barcode and the number of labels you require.
It will print labels off with the barcode and text Control.

to print ad hoc labels:

visit /print_labels/ad_hoc_plate

scan or type barcode and text and it wil print off a single label the same as destination plates but with the text of your choice.

File: pages/print_labels/ad_hoc_plate.vue [hunk 1]
```diff
@@ -0,0 +1,126 @@
+<template>
+  <b-container>
+    <b-row>
+      <b-col>
+        <h1>Print Control plate labels</h1>
+        <p class="lead"></p>
+
+        <!-- TODO: better in a component of its own? -->
```
Comment (harrietc52) [inline pages/print_labels/ad_hoc_plate.vue:8]: There is an Alert component, which might be useful instead. For an example it's being implemented in the Beckman Cherrypick view

File: pages/print_labels/ad_hoc_plate.vue [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+          </b-alert>
+          <b-alert :show="isSuccess" dismissible variant="success">
+            {{ alertMessage }}
+          </b-alert>
+          <b-alert :show="isBusy" dismissible variant="warning">
+            {{ alertMessage }}
+          </b-alert>
+        </p>
+        <p>
+          <label for="selectPrinter">
```
Comment (harrietc52) [inline pages/print_labels/ad_hoc_plate.vue:21]: A Printer Select component would be a good one to abstract, think it is being used in a few places

**Review by harrietc52 — APPROVED** [review]
All good! Agree with the TODOs.
One way to refactor, as the Destination Plate/ Ad Hoc/ Control Plate/ Source Plate pages are all quite similar, could be having a Print Label page as a template for all these pages, then using slots, the duplication could hopefully be reduced. 
Lots of tests (which made the PR look massive but luckily is isn't!) but like you say, they're for code coverage and can go once refactored.
Nice mocking in the tests too, you've really got the hang of that.
Story card also has the 'On Hold' label, not sure if thats needed or whether, becuase this has already been deployed to UAT, it can move straight to the 'Ready to Deploy to PROD' column

**Review by harrietc52 — APPROVED** [review]
All looks fine, a big refactor is needed/ would be nice to do!
We think the 4 print label pages are similar enough to have in one component, with some logic/ config determining the differences. That would remove the need to refactor out the status/ printer component/ navbar etc

File: components/PrintLabelsRouter.vue [hunk 3]
```diff
@@ -0,0 +1,28 @@
+<template>
+  <b-navbar
+    ref="navbar"
+    toggleable="md"
+    type="light"
+    class="bg-white"
+    variant="info"
+  >
+    <b-navbar-nav class="ml-auto">
+      <b-nav-item to="/print_labels/destination_plates"
```
[the remark this answers is not in the document]
Comment (stevieing) [reply components/PrintLabelsRouter.vue:10]: I added destination plate labels to the index so when people go to this page it doesn't change. Most of the time they will be printing off destination plate labels.

File: pages/print_labels/index.vue [hunk 4]
```diff
@@ -0,0 +1,21 @@
+<template>
+  <b-container>
+    <b-row>
+      <b-col>
+        <DestinationPlates />
```
[the remark this answers is not in the document]
Comment (stevieing) [reply pages/print_labels/index.vue:5]: I tried that but it didn't seem to work as expected.

File: pages/print_labels/ad_hoc_plate.vue [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+import Sprint from '@/modules/sprint'
+import config from '@/nuxt.config'
+import PrintLabelsRouter from '@/components/PrintLabelsRouter'
+
+export default {
+  components: {
+    PrintLabelsRouter
+  },
+  props: {
+    printers: {
```
[the remark this answers is not in the document]
Comment (stevieing) [reply pages/print_labels/ad_hoc_plate.vue:76]: I always thought it was standard practise to have mutable as data and immutable as props?

File: pages/print_labels/ad_hoc_plate.vue [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+      type: Array,
+      default() {
+        return config.publicRuntimeConfig.printers.split(',')
```
[the remark this answers is not in the document]
Comment (stevieing) [reply pages/print_labels/ad_hoc_plate.vue:79]: Can you have lists in environment variables? I thought I had tried it and they all get converted to a string.

**Pull request merged**